src/Microsoft.AspNet.Cors/CorsMiddleware.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Cors.Core;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.WebUtilities;
+
+namespace Microsoft.AspNet.Cors
+{
+    /// <summary>
+    /// An ASP.NET middleware for handling CORS.
+    /// </summary>
+    public class CorsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly CorsOptions _options;
+
+        /// <summary>
+        /// Instantiates a new <see cref="CorsMiddleware"/>.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        public CorsMiddleware(
+            [NotNull]RequestDelegate next,
+            [NotNull]CorsOptions options)
+        {
+            _next = next;
+            _options = options;
+
+            _options.PolicyProvider = _options.PolicyProvider ?? new CorsPolicyProvider();
+            _options.CorsEngine = _options.CorsEngine ?? new CorsEngine();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var corsRequestContext = new CorsRequestContext(context);
+
+            if (corsRequestContext.IsCorsRequest)
```

Comment:
Create a new instance of context to check if a request is a Cors request looks like an overkill, since it only takes reading a header. We can make the `IsCorsRequest` a static method. If the request is indeed a Cors request then create the `CorsRequestContext`. It avoid unnecessary allocation.